File: LineLexer.java
public class LineLexer {
    private int columnNumber;
    private final String line;
    private final int lineNumber;

    public LineLexer(String line, int lineNumber) {
        this.line = line;
        this.lineNumber = lineNumber;
        this.columnNumber = 0;
    }

    public Token nextToken() {
        while (Character.isWhitespace(getCurrentChar())) {
            columnNumber++;
        }
        if (getCurrentChar() == 0) {
            return new Token(Token.Kind.EOL, "<eol>", lineNumber, line.length() - 1);
        }

        int column = columnNumber;

        // Identifiers and keywords
        if (isIdentifierStart(getCurrentChar())) {
            StringBuilder sb = new StringBuilder();
            do {
                sb.append(getCurrentChar());
                columnNumber++;
            } while (isIdentifierPart(getCurrentChar()));
            Token.Kind kind = Token.lookupKeyword(sb.toString());
            if (kind == null) {
                kind = Token.Kind.IDENTIFIER;
            }
            return new Token(kind, sb.toString(), lineNumber, column);
        }

        // Integer literals
        if (getCurrentChar() == '0') {
            columnNumber++;
            return new Token(Token.Kind.INTEGER_LITERAL, "0", lineNumber, column);
        }
        if (Character.isDigit(getCurrentChar())) {
            StringBuilder sb = new StringBuilder();
            do {
                sb.append(getCurrentChar());
                columnNumber++;
            } while (Character.isDigit(getCurrentChar()));
            return new Token(Token.Kind.INTEGER_LITERAL, sb.toString(), lineNumber, column);
        }

        // String literals
        if (getCurrentChar() == '"') {
            columnNumber++; // Skip opening quote
            StringBuilder sb = new StringBuilder();
            while (getCurrentChar() != '"' && getCurrentChar() != 0) {
                if (getCurrentChar() == '\n') {
                    return new Token(Token.Kind.ERROR,
                        "line end in string literal", lineNumber, columnNumber);
                }
                sb.append(getCurrentChar());
                columnNumber++;
            }
            if (getCurrentChar() != '"') {
                return new Token(Token.Kind.ERROR,
                    "unclosed string literal", lineNumber, columnNumber);
            }
            columnNumber++; // Skip closing quote
            return new Token(Token.Kind.STRING_LITERAL,
                sb.toString(), lineNumber, columnNumber);
        }

        String text = Character.toString(getCurrentChar());
        Token.Kind kind = Token.lookupSymbol(getCurrentChar());
        columnNumber++;
        if (kind == null) {
            return new Token(Token.Kind.ERROR,
                "invalid character: '" + text + "'", lineNumber, column);
        }
        return new Token(kind, text, lineNumber, column);
    }

    private char getCurrentChar() {
        return (columnNumber < line.length()) ? line.charAt(columnNumber) : 0;
    }

    private boolean isIdentifierStart(char c) {
        return Character.isLetter(c) || c == '_' || c == '$';
    }

    private boolean isIdentifierPart(char c) {
        return isIdentifierStart(c) || Character.isDigit(c);
    }
}

File: Interpreter/LineLexer.cs
using System.Text;

namespace Basic
{
    public class LineLexer
    {
        private const char Eol = '\0';

        private readonly string _line;
        private int _lineNumber;
        private int _columnNumber;

        public LineLexer(string line, int lineNumber)
        {
            _line = line;
            _lineNumber = lineNumber;
        }

        public Token NextToken()
        {
            while (char.IsWhiteSpace(CurrentChar))
            {
                Advance();
            }

            if (CurrentChar == '/')
            {
                Advance();
                if (CurrentChar == '/')
                {
                    _columnNumber = _line.Length;
                }
                else
                {
                    return new Token(TokenType.Div, "/", _lineNumber, _columnNumber - 1);
                }
            }

            if (CurrentChar == Eol)
            {
                return new Token(TokenType.Eol, "<eol>", _lineNumber, _line.Length);
            }

            if (IsIdentifierStart(CurrentChar))
            {
                return LexIdentifierOrKeyword();
            }

            if (char.IsDigit(CurrentChar))
            {
                return LexIntegerLiteral();
            }

            if (CurrentChar == '"')
            {
                return LexStringLiteral();
            }

            TokenType type = Token.LookupSymbol(CurrentChar) ?? TokenType.Error;
            string text = char.ToString(CurrentChar);
            
            if (type == TokenType.Error)
            {
                text = $"character '{text}' is not allowed here";
            }

            Advance();

            return new Token(type, text, _lineNumber, _columnNumber - 1);
        }

        private Token LexIdentifierOrKeyword()
        {
            int column = _columnNumber;
            var text = new StringBuilder();

            while (IsIdentifierPart(CurrentChar))
            {
                text.Append(CurrentChar);
                Advance();
            }

            TokenType type = Token.LookupKeyword(text.ToString()) ?? TokenType.Identifier;

            return new Token(type, text.ToString(), _lineNumber, column);
        }

        private Token LexIntegerLiteral()
        {
            int column = _columnNumber;
            var numeral = new StringBuilder();

            while (char.IsDigit(CurrentChar))
            {
                numeral.Append(CurrentChar);
                Advance();
            }

            return new Token(TokenType.IntegerLiteral, numeral.ToString(), _lineNumber, column);
        }

        private Token LexStringLiteral()
        {
            int column = _columnNumber;

            Advance(); // Skip opening quote

            var text = new StringBuilder();

            while (CurrentChar != '"' && CurrentChar != Eol)
            {
                if (CurrentChar == '\n')
                {
                    string err = "line end in string literal";
                    return new Token(TokenType.Error, err, _lineNumber, column);
                }

                text.Append(CurrentChar);
                Advance();
            }

            if (CurrentChar != '"')
            {
                string err = "unclosed string literal";
                return new Token(TokenType.Error, err, _lineNumber, column);
            }

            Advance(); // Skip closing quote

            return new Token(TokenType.StringLiteral, text.ToString(), _lineNumber, column);
        }

        private char CurrentChar
            => (_columnNumber < _line.Length) ? _line[_columnNumber] : Eol;
        
        private void Advance()
        {
            _columnNumber++;
        }

        private static bool IsIdentifierStart(char c)
            => char.IsLetter(c) || c == '_' || c == '$';
        
        private static bool IsIdentifierPart(char c)
            => IsIdentifierStart(c) || char.IsDigit(c);
    }
}